include/ibamr/HierarchyTimeInterpolator.h

```diff
@@ -0,0 +1,182 @@
+
+#ifndef included_IBAMR_INSStaggeredMeanFlowCalculator
+#define included_IBAMR_INSStaggeredMeanFlowCalculator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibamr/config.h>
+
+#include <ibamr/SnapshotCache.h>
+
+#include "CellVariable.h"
+#include "IntVector.h"
+#include "SideVariable.h"
+#include "tbox/Database.h"
+#include "tbox/Pointer.h"
+
+#include <string>
+#include <vector>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class HierarchyTimeInterpolator provides a method of determining and storing average fields over periodic
+ * intervals. This class can also be used to determine intermediate velocity fields when at a periodic steady state.
+ */
+class HierarchyTimeInterpolator
+{
+public:
+    /*!
+     * The constructor for class HierarchyTimeInterpolator sets some default values and determines data centering. The
+     * expected period and number of snapshots must be set in the input database. This class assumes the snapshots will
+     * be taken at equidistant points along the periodic interval.
+     */
+    HierarchyTimeInterpolator(std::string object_name,
+                              SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy);
+
+    /*!
+     * The constructor for class HierarchyTimeInterpolator sets some default values and determines data centering. In
+     * this constructor, the times at which snapshots are taken are set, as well as the period length.
+     */
+    HierarchyTimeInterpolator(std::string object_name,
+                              std::set<double> snapshot_time_points,
+                              double period,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy);
+
+    /*!
+     * \brief The destructor for class HierarchyTimeInterpolator deallocates patch data as needed.
+     */
+    ~HierarchyTimeInterpolator();
+
+    /*!
+     * Clear the snapshots taken with this class.
+     */
+    void clearSnapshots();
+
+    /*!
+     * Fill a patch index at a particular time point. This will perform a time interpolation of the snapshots stored in
+     * the SnapshotCache object. Note if there is only one snapshot (i.e. the period is 0 and we are assumed to be at a
+     * steady state), a copy will be performed
+     */
+    void
+    fillSnapshotAtTime(int u_idx, double time, SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy);
+
+    /*!
+     * Update the time averaged snapshot. The update is computed through the relation
+     * ``` u_avg = u_avg + 1/N * (u - u_avg)```
```

Comment:
state what N is here?

Reply:
Also, this averaging process is also useful for models that are not periodic but instead give a steady mean value. Can/should we support both in the same class?

Reply:
This class should take care of that. You can store one snapshot and set the "period" to be 0.